File: app/src/main/java/com/asteph11/finalproject/models/viewmodels/MatchViewModel.java
package com.asteph11.finalproject.models.viewmodels;

import android.util.Log;

import androidx.lifecycle.ViewModel;

import com.asteph11.finalproject.models.data.Player;
import com.asteph11.finalproject.models.data.TurnHandler;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QueryDocumentSnapshot;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class MatchViewModel extends ViewModel {

    private ArrayList<String> topnames;
    private FirebaseFirestore db;
    private TurnHandler turnHandler;

    public void init(FirebaseFirestore db) {
        this.db = db;
        if(turnHandler == null) {
            initMatch(new Player("NA"), new Player("NA"));
        }
        if(topnames == null) {
            setTopList();
        }
    }

    public void initMatch(Player p1, Player p2) {
        turnHandler = new TurnHandler(p1, p2);
    }

    public Player getP1() {
        return turnHandler.getP1();
    }

    public Player getP2() {
        return turnHandler.getP2();
    }

    public TurnHandler getTurnHandler() {
        return turnHandler;
    }

    public boolean playersReady() throws Exception {
        if(turnHandler == null) {
            throw new Exception("TurnHandler is null!");
        }
        return turnHandler.isReady();
    }

    public boolean isGameFinished() throws Exception {
        return turnHandler.isFinished();
    }

    public void addWinner() {
        db.collection("Winners")
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        Map<String, String> user = new HashMap<>();
                        user.put("name", turnHandler.getCurrentPlayer().getName());
                        user.put("timestamp", ""+System.currentTimeMillis());
                        db.collection("Winners")
                                .document(turnHandler.getCurrentPlayer().getName()).set(user);
                    } else {
                        Log.w("FBDB", "Error getting documents.", task.getException());
                    }
                });
    }

    public ArrayList<String> setTopList() {
        db.collection("Winners").orderBy("timestamp", Query.Direction.DESCENDING)
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        if(topnames == null) {
                            topnames = new ArrayList<>();
                        }
                        for (QueryDocumentSnapshot document : task.getResult()) {
                            String name = document.getData().get("name") + "";
                            topnames.add(name);
                            Log.d("Names", topnames.get(topnames.size()-1));
                        }
                    } else {
                        Log.w("FBDB", "Error getting documents.", task.getException());
                    }
                });

        return topnames;
    }

    public ArrayList<String> getTopList() {
        return topnames;
    }


    public void reset() {
        turnHandler = null;
    }
}

File: app/src/main/java/com/asteph11/finalproject/models/viewmodels/MatchViewModel.kt
package com.asteph11.finalproject.models.viewmodels

import android.util.Log
import androidx.lifecycle.ViewModel
import com.asteph11.finalproject.models.data.Player
import com.asteph11.finalproject.models.data.TurnHandler
import com.google.android.gms.tasks.Task
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.QuerySnapshot

class MatchViewModel : ViewModel() {
    var topList: ArrayList<String>? = null
        private set
    private var db: FirebaseFirestore? = null
    var turnHandler: TurnHandler? = null
        private set

    fun init(db: FirebaseFirestore?) {
        this.db = db
        if (turnHandler == null) {
            initMatch(Player("NA"), Player("NA"))
        }
        if (topList == null) {
            setTopList()
        }
    }

    fun initMatch(p1: Player?, p2: Player?) {
        turnHandler = TurnHandler(p1, p2)
    }

    val p1: Player?
        get() = turnHandler!!.p1
    val p2: Player?
        get() = turnHandler!!.p2

    @Throws(Exception::class)
    fun playersReady(): Boolean {
        if (turnHandler == null) {
            throw Exception("TurnHandler is null!")
        }
        return turnHandler!!.isReady
    }

    fun addWinner() {
        db!!.collection("Winners")
            .get()
            .addOnCompleteListener { task: Task<QuerySnapshot?> ->
                if (task.isSuccessful) {
                    val user: MutableMap<String, String> = HashMap()
                    user["name"] = turnHandler!!.currentPlayer!!.name
                    user["timestamp"] = "" + System.currentTimeMillis()
                    db!!.collection("Winners")
                        .document(turnHandler!!.currentPlayer!!.name).set(user)
                } else {
                    Log.w("FBDB", "Error getting documents.", task.exception)
                }
            }
    }

    fun setTopList(): ArrayList<String>? {
        db!!.collection("Winners").orderBy("timestamp", Query.Direction.DESCENDING)
            .get()
            .addOnCompleteListener { task: Task<QuerySnapshot> ->
                if (task.isSuccessful) {
                    if (topList == null) {
                        topList = ArrayList()
                    }
                    for (document in task.result) {
                        val name = document.data["name"].toString() + ""
                        topList!!.add(name)
                        Log.d("Names", topList!![topList!!.size - 1])
                    }
                } else {
                    Log.w("FBDB", "Error getting documents.", task.exception)
                }
            }
        return topList
    }

    fun reset() {
        turnHandler = null
    }
}